scripts/run_portserver.py

```diff
@@ -0,0 +1,491 @@
+# Copyright 2020 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Adapted from portserver.py in the python_portpicker project (commit
+# bf6250b121cd610c3280cd748b5702a2be040f41) at
+# https://github.com/google/python_portpicker/blob/main/src/portserver.py,
+# which is Copyright 2015 Google Inc. All Rights Reserved. The following
+# changes were made by The Oppia Authors to this file when it was
+# originally introduced. More may have been made since.
+#
+# * We used threading instead of asyncio for compatibility with Python 2.
+#   This change involved adding a Server class to replace asyncio's
+#   event loop.
+# * We re-wrote the logger format string to be compatible with Python 2.
+#   We kept the same general string structure.
+# * We renamed _handle_port_request to handle_port_request and removed
+#   the original handle_port_request function from the
+#   _PortServerRequestHandler class. We also made _handle_port_request
+#   return the request response instead of writing it to a provided
+#   writer.
+# * We re-wrote the main function to use the Server class instead of
+#   asyncio. In doing so we also simplified the main function by
+#   removing the signal.SIGUSR1 handler.
+# * We changed the default portserver address to `portserver.sock` to
+#   avoid problems introduced by using an address beginning with a null
+#   byte.
+
+"""A server to hand out network ports to applications running on one host.
+Typical usage:
+ 1) Run one instance of this process on each of your unittest farm hosts.
+ 2) Set the PORTSERVER_ADDRESS environment variable in your test runner
+    environment to let the portpicker library know to use a port server
+    rather than attempt to find ports on its own.
+$ /path/to/portserver.py &
+$ export PORTSERVER_ADDRESS=portserver.sock
+$ # ... launch a bunch of tests that use portpicker ...
+"""
+
+from __future__ import absolute_import  # pylint: disable=import-only-modules
+from __future__ import unicode_literals  # pylint: disable=import-only-modules
+
+import argparse
+import collections
+import logging
+import os
+import socket
+import sys
+import threading
+
+import python_utils
+
+_PROTOCOLS = [(socket.SOCK_STREAM, socket.IPPROTO_TCP),
+              (socket.SOCK_DGRAM, socket.IPPROTO_UDP)]
+
+
+def _get_process_command_line(pid):
+    """Get the command for a process."""
```

Comment:
Add Args/Return etc., here and elsewhere, using the normal format.

Reply:
done in 97bba1a74